test/unified_format.py

```diff
@@ -0,0 +1,997 @@
+
+    def _testOperation_assertIndexNotExists(self, spec):
+        collection = self.client[spec['databaseName']][spec['collectionName']]
+        for index in collection.list_indexes():
+            self.assertNotEqual(spec['indexName'], index['name'])
+
+    def run_special_operation(self, spec):
+        opname = spec['name']
+        method_name = '_testOperation_%s' % (opname,)
+        try:
+            method = getattr(self, method_name)
+        except AttributeError:
+            self.fail('Unsupported special test operation %s' % (opname,))
+        else:
+            method(spec['arguments'])
+
+    def run_operations(self, spec):
+        for op in spec:
+            target = op['object']
+            if target != 'testRunner':
+                self.run_entity_operation(op)
+            else:
+                self.run_special_operation(op)
+
+    def check_events(self, spec):
+        for event_spec in spec:
+            client_name = event_spec['client']
+            events = event_spec['events']
+            listener = self.entity_map.get_listener_for_client(client_name)
+
+            if len(events) == 0:
+                self.assertEqual(listener.results, [])
+                continue
+
+            if len(events) > len(listener.results):
+                self.fail('Expected to see %s events, got %s' % (
+                    len(events), len(listener.results)))
+
+            for idx, expected_event in enumerate(events):
+                self.match_evaluator.match_event(
+                    expected_event, listener.results[idx])
+
+    def verify_outcome(self, spec):
+        for collection_data in spec:
+            coll_name = collection_data['collectionName']
+            db_name = collection_data['databaseName']
+            expected_documents = collection_data['documents']
+
+            coll = self.client.get_database(db_name).get_collection(
+                coll_name,
+                read_preference=ReadPreference.PRIMARY,
+                read_concern=ReadConcern(level='local'))
+
+            if expected_documents:
+                sorted_expected_documents = sorted(
+                    expected_documents, key=lambda doc: doc['_id'])
+                actual_documents = list(
+                    coll.find({}, sort=[('_id', ASCENDING)]))
+                self.assertListEqual(sorted_expected_documents,
+                                     actual_documents)
+
+    def run_scenario(self, spec):
+        # process createEntities
+        self.entity_map = EntityMapUtil(self)
+        self.entity_map.create_entities_from_spec(
+            self.TEST_SPEC.get('createEntities', []))
+
+        # process initialData
+        self.insert_initial_data(self.TEST_SPEC.get('initialData', []))
+
+        # process test-level runOnRequirements
+        run_on_spec = spec.get('runOnRequirements', [])
+        if not self.should_run_on(run_on_spec):
+            raise unittest.SkipTest('runOnRequirements not satisfied')
```

Comment:
Shouldn't runOnRequirements happen first, ie before processing `createEntities` and `initialData`?

Reply:
good point. Also moved up the processing of skipTest.